### DISCUSSION TITLE
Shrinking down executable image

### DISCUSSION THREAD

**Discussion by aalmiray** [oracle/graal#4184] [Q&A]
I'm currently looking at options to generate a native image executable for the [jreleaser/jreleaser](https://github.com/jreleaser/jreleaser/) CLI. Current settings create an image close to 100M in size and I wonder what could be done to shrink it down.

All classes found int he `org.jreleaser.*` packages should be included in the image. There are a few places where proxy and reflection config must be declared. Most of the codebase is reflection free but external configuration is read through Jackson (hence reflection), some network services require reflection/proxies as well (Feign, Hub4j, Twitter4j).

I used the agent to gather information and ended up manually updating the generated files into https://github.com/jreleaser/jreleaser/tree/graal/apps/jreleaser/src/main/resources/META-INF/native-image/org.jreleaser/jreleaser
This manual update is likely responsible for the growth in size.

Additionally every project generates reflect and proxy config files using custom APT. See attached ZIP with all generated files
[native-image.zip](https://github.com/oracle/graal/files/7822258/native-image.zip)

The executable is created passing these additional command flags:

```
-H:Name=jreleaser \
-Duser.language=en -H:IncludeLocales=en,ca,de,es,fr,hi,it,ja,nl,pt_BR,zh_TW,ru \
-H:Optimize=2 -H:+RemoveUnusedSymbols
```

**Comment by neomatrix369** [comment]
Hey @aalmiray I have a couple of resources for you to look at: 

- [Compressed GraalVM Native Images: the best startup for Java apps comes in tiny packages](https://medium.com/graalvm/compressed-graalvm-native-images-4d233766a214?source=collection_home---6------12-----------------------)
- [UPX to compress GraalVM Images](https://twitter.com/juliendubois/status/1337005381436977152) - in the thread you can see I have shared a working example of how to compress a native-image using UPX

You can also find other native-image and other GraalVM related posts here https://github.com/neomatrix369/awesome-graal/blob/master/blog-posts.md 


I hope this helps - maybe you wish to accomplish this in another way, but UPX seems to be the most portable and one of the widely tested ways to get around this. It is of course after the fact the native-image builder creates the native-image, and does not solve the issue at build/image creation level.

**Reply by neomatrix369** [reply]
Here's the tweet with the example: https://twitter.com/theNeomatrix369/status/1340607120778145792

If you adapt the shellscript with your project, you can do all the CLI actions related to GraalVM using `--option` switches

So far I have see upwards of 30% compression - this of course varies from binary to binary.

**Reply by fniephaus** [reply]
Ha, you beat me to it @neomatrix369! However, I think it's a bit too early to apply UPX. Instead, I'd suggest to look at the cause, which seems to be the newly introduced reflection configuration. Including all fields, methods, and constructors may not be necessary to create a working binary.

**Reply by neomatrix369** [reply]
Yes that's true, but UPX can be applied only after the fact -- i.e. after the image has been created.

Good suggestion about the steps you suggested.

**Reply by neomatrix369** [reply]
Also, I think somehow the UPX step can be made part of the native-image build process - via an additional switch for those who wish to go that path, saves them all the extra steps they would need to do - my example programs are available for everyone to reuse.

**Reply by aalmiray** [reply]
Hi @neomatrix36. Thanks. I'm aware of UPX and @fniephaus mentioned it's a cure for the symptoms but not for the actual cause, which is why this discussion thread started.

**Reply by neomatrix369** [reply]
Ok thats fine you are looking to solve it at the internal levels.

Even after this issue is fixed I feel, the image can be further reduced in size using UPX, which either can be done externally or be built into the native-image builder.

**Reply by sergeevabc** [reply]
Binaries generated by GraalVM 17-24 and compressed with UPX [do not work](https://github.com/upx/upx/issues/670#issuecomment-2869161347) without a hack, alas.

**Comment by fniephaus** [comment]
Thanks for opening the discussion, @aalmiray. What's the easiest way to reproduce such a build? I'd like to run it through the [Dashboard](https://www.graalvm.org/dashboard/?ojr=help%3Btopic%3Dgetting-started.md).

**Reply by neomatrix369** [reply]
@fniephaus Dashboard looks great, did you see this share from someone who has data from GraalVM published on the ELK stack and even on prometheus? If I come across it I will share it here.

**Reply by aalmiray** [reply]
These are the steps required to make such a build:

### Requirements
* Java 11. **Cannot be 16 or greater**
* You may use Graal 11, that would help reducing 1 config step.

Building a local JReleaser distribution:
```
git clone https://github.com/jreleaser/jreleaser.git
cd jreleaser
git checkout graal -t origin/graal
./gradlew -Prelease=true :jreleaser:distZip
```

This creates the distribution at `apps/jreleaser/build/distributions/jreleaser-1.0.0-SNAPSHOT.zip`. Unpack in a directory of your choosing then invoke the `jreleaser` command:

```
unzip -d. apps/jreleaser/build/distributions/jreleaser-1.0.0-SNAPSHOT.zip
./jreleaser-1.0.0-SNAPSHOT/bin/jreleaser assemble -a native-image
```

If successful you'll see an executable placed in `out/jreleaser/assemble/jreleaser-native/native-image/jreleaser`
Full log is available at `out/jreleaser/trace.log`.

**Reply by fniephaus** [reply]
> did you see this share from someone who has data from GraalVM published on the ELK stack and even on prometheus?

@neomatrix369 no, I did not. please share!

> These are the steps required to make such a build

@aalmiray Thanks, will take a look later today.

**Reply by fniephaus** [reply]
I'm running into the following error and don't know how to proceed. Also, is there any way to get closer to the native-image invocation and how can I regenerate the reflection config with the agent?

```
$ JRELEASER_PROJECT_VERSION=1.0.0-SNAPSHOT ./jreleaser-1.0.0-SNAPSHOT/bin/jreleaser assemble -a native-image
[INFO]  JReleaser 1.0.0-SNAPSHOT
[INFO]  Configuring with jreleaser.yml
[INFO]    - basedir set to /home/fniephaus/dev/tmp/jreleaser
[INFO]  Reading configuration
[INFO]  Loading variables from /home/fniephaus/.jreleaser/config.properties
[WARN]  Variables source /home/fniephaus/.jreleaser/config.properties does not exist
[INFO]  Validating configuration
[INFO]  Project version set to 1.0.0-SNAPSHOT
[INFO]  Release is snapshot
[INFO]  Timestamp is 2022-01-06T18:19:27.488344+01:00
[INFO]  HEAD is at baae46e
[INFO]  Platform is linux-x86_64
[INFO]  dryrun set to false
[INFO]  Included assemblers: [native-image]
[INFO]  Assembling all distributions with native-image
[INFO]    [native-image] assembling jreleaser-native distribution
[INFO]    [native-image] - jreleaser-early-access-linux-x86_64
[INFO]    [native-image] [jreleaser:31041]    classlist:     716.90 ms,  0.96 GB
[INFO]    [native-image] [jreleaser:31041]      [total]:     745.79 ms,  0.96 GB
[ERROR]   [native-image] Fatal error:java.lang.NoClassDefFoundError: picocli/CommandLine
[ERROR]   [native-image] 	at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
[ERROR]   [native-image] 	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3166)
[ERROR]   [native-image] 	at java.base/java.lang.Class.getDeclaredMethod(Class.java:2473)
[ERROR]   [native-image] 	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:355)
[ERROR]   [native-image] 	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:569)
[ERROR]   [native-image] 	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:122)
[ERROR]   [native-image] 	at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:599)
[ERROR]   [native-image] Caused by: java.lang.ClassNotFoundException: picocli.CommandLine
[ERROR]   [native-image] 	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:476)
[ERROR]   [native-image] 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
[ERROR]   [native-image] 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
[ERROR]   [native-image] 	... 7 more
[ERROR]   [native-image] Error: Image build request failed with exit status 1
[INFO]  Writing output properties to out/jreleaser/output.properties
[ERROR] JReleaser failed after 1.314 s
```

**Reply by aalmiray** [reply]
Ah, I think I missed a command. This problem is caused by missing JARs in the classpath. Can be fixed by invoking this command

```
./gradlew -Prelease=true :jreleaser:distZip :jreleaser:assemble
```

**Reply by fniephaus** [reply]
Ok, I can now build jreleaser. It's 113M on Linux right now. How do I run the tests to ensure the binary is functional? And how did you generate the reflection config? I'd like to regenerate the file.

**Reply by aalmiray** [reply]
The easiest way is to test it against a real repository.

1. Fork https://github.com/aalmiray/app/
2. Clone the fork to your local environment
3. Create a GitHub Personal Access token with read/write access https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token
4. Store the token in an env var named JRELEASER_GITHUB_TOKEN
5. Build the project binaries with `./mvnw verify`
6. Release it with `jreleaser release`

I used the agent to capture the reflect config file.

**Reply by fniephaus** [reply]
I see. I'm afraid I don't want to create a PAT with read/write access for my entire GH account without reviewing what this access is used for and right now, I don't have the time for such a review. I'm sure it's not supposed to do anything dangerous but I like to be cautious with things like this. Is there any other way to test the binary apart from creating a dummy GH account for this?

**Reply by aalmiray** [reply]
You can pass any value as the github token but the release process will eventually fail. You can also invoke the release in dryrun mode by passing the `-y` command flag, but that of course will exercise less behavior. There's no other way to tell if the binary actually works or not unless it talks to real services 🤷‍♂️

**Reply by fniephaus** [reply]
I see that you built some [infrastructure for generating the reflection config](https://github.com/jreleaser/jreleaser/blob/b6a2f27e09af01d17507b61e7986dc0695a433b1/infra/nativeimage-processor/src/main/java/org/jreleaser/infra/nativeimage/processor/ReflectConfigGenerator.java#L59-L83). How is this supposed to be used? It does not seem to integrate the NI agent, does it?

**Reply by aalmiray** [reply]
It does not. It's an annotation processor triggered on every build. The only manually created files are located inside the `apps/jreleaser/src/main/resources/META-INF` directory. Every other reflect/proxy config files found in the final artifacts are created using the annotation processor. You can see them on every module at `build/classes/java/main/META-INF`.

**Reply by fniephaus** [reply]
Ok, so in general, I believe it's not a good idea to force-include all public methods, fields, and constructors. Instead, the reachability analysis should be able to find out what's needed and what's not. If it doesn't, then there's something wrong with either the setup or the analysis itself.
Looking at the number of dependencies, I'm not surprised that the binary ends up being quite large. Nonetheless, it may be possible to generate much smaller binaries that are functional.

So again, how did you use the agent to create the reflect-config,json files in your project? Did you run it as part of a releasing process with your [demo app](https://github.com/aalmiray/app/)? Also, are there any development dependencies that may accidentally be included in the binary?

**Reply by aalmiray** [reply]
There are no development dependencies included in the resulting binary nor Java distro.

I ran the agent configuring `JAVA_OPTS` per the guide instructions when running the binary as a Java distribution.

I'm aware that the agent defines a smaller set of reflection elements but I wasn't sure all codepaths were exercised during live tests hence why a wider set of methods/constructors were manually defined in the resulting reflect config file found inside `apps/jreleaser`.

OTOH the automatically generated reflect config files must remain as they are.

**Reply by fniephaus** [reply]
So with your current setup, you can create a functional +100MB jreleaser binary that you could potentially ship at some point, right? And you can shrink it down to something like 30MB using UPX, correct? If so, I will do some more experiments when I have some time or are you currently blocked because of this?

In summary, I think marking too many things as reachable by hand can certainly blow up image sizes. However, it seems this is the only way you found to generate a working binary, so maybe jreleaser is a good example for where NI tooling support could be improved.

BTW: Are you aware of the [Gradle plugin for GraalVM Native Image building](https://graalvm.github.io/native-build-tools/latest/gradle-plugin.html)? If you could give that a try and report your experience that'd be great. Maybe it even improves image sizes! :)

**Reply by aalmiray** [reply]
Yes, it seems that the current setup may work. Now that I've figure out how to run UPX as part of the build pipeline for all hosted runners on GitHub I can build with more confidence. 

The Gradle native plugin could be used instead of delegating to JReleaser's assemble which wraps the native-image executable, but only if it offers more than just using the native-image executable, does it?

The current explicit reflect config is quite wide in its reach as an overabundance of caution that some codepaths were not exercised with the agent. I'd rather have a big binary than a broken binary. Time will tell how that size can be trimmed down with more accurate configuration. 

Thanks.
